shared/src/econ/mod.rs

```diff
@@ -253,6 +262,36 @@ where
     }
 }
 
+macro_rules! impl_mul {
+    ($(impl<$($a:lifetime),*> Mul<$Other:ty> for $Self:ty;)*) => {$(
+        impl<$($a),*> Mul<$Other> for $Self {
+            type Output = TokenAmount;
+
+            #[inline]
+            fn mul(self, other: $Other) -> TokenAmount {
+                other * self
+            }
+        }
+    )*}
+}
+
+macro_rules! impl_muls {
+    ($($t:ty,)*) => {$(
+        impl_mul! {
+            impl<> Mul<TokenAmount> for $t;
+            impl<'b> Mul<&'b TokenAmount> for $t;
+            impl<'a> Mul<TokenAmount> for &'a $t;
+            impl<'a, 'b> Mul<&'b TokenAmount> for &'a $t;
+        }
+    )*};
+}
+
+impl_muls! {
+    u8, u16, u32, u64, u128,
+    i8, i16, i32, i64, i128,
+    BigInt,
+}
```

Comment:
We had already implemented the left-hand size (`TokenAmount * {integer}`), this implements the right-hand side (`TokenAmount * {integer}`). If you're wondering _why_ we have to do this... reasons.

Reply:
Expands to:


```rust
impl Mul<TokenAmount> for u8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for u8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a u8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a u8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for u16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for u16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a u16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a u16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for u32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for u32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a u32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a u32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for u64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for u64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a u64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a u64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for u128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for u128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a u128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a u128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for i8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for i8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a i8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a i8 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for i16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for i16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a i16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a i16 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for i32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for i32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a i32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a i32 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for i64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for i64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a i64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a i64 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for i128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for i128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a i128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a i128 {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl Mul<TokenAmount> for BigInt {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'b> Mul<&'b TokenAmount> for BigInt {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a> Mul<TokenAmount> for &'a BigInt {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: TokenAmount) -> TokenAmount {
        other * self
    }
}
impl<'a, 'b> Mul<&'b TokenAmount> for &'a BigInt {
    type Output = TokenAmount;
    #[inline]
    fn mul(self, other: &'b TokenAmount) -> TokenAmount {
        other * self
    }
}
```